File: src/lib/net.js
/** ZeroMQ export helpers for client/server nodes */

const zeromq = require("zeromq");
const util = require('./util');
const log = util.logpre('zmq');
const os = require('os');

const { Dealer, Router } = zeromq;
const { args, json } = util;
const proto = "tcp";
const settings = {
  heartbeat: 1000,
  dead_client: 5000
};

function zmq_settings(rec) {
  Object.assign(settings, rec);
}

function zmq_server(port, onmsg, opt = { sync: false }) {
  const sock = new Router;
  const cidmap = {};
  
  // send a message to a known client-id endpoint
  // because 0MQ, the endpoint could be down
  // thus heartbeating is required by users of the class
  function send(cid, msg) {
    const id = cidmap[cid];
    if (id) {
      sock.send([ id, json(msg) ]);
        return true;
    } else {
        return false; // dead endpoint
    }
  }

  // remove dead endpoint record
  // server wrappers like proxy are responsible for heartbeating
  function remove(cid) {
    delete cidmap[cid];
  }

  // start a background async message receiver
  (async function () {
    await sock.bind(`${proto}://*:${port}`);
    log({ listening: proto, port, opt }); 
  
    for await (const [id, msg] of sock) {
      const cid = id.readUInt32BE(1).toString(36).toUpperCase();
      cidmap[cid] = id;
      const req = JSON.parse(msg);
      const rep = await onmsg(req, cid, send);
      if (rep !== undefined) {
        const pro = sock.send([id, json(rep)]);
        if (opt.sync) await pro;
      }
    }
  }());
  
  return { send, remove };
}

function zmq_client(host = "127.0.0.1", port) {
  const sock = new Dealer;
  sock.connect(`${proto}://${host}:${port}`);
  log({ connected: `${host}:${port}` });

  async function send(request) {
    await sock.send(json(request));
  }

  async function recv() {
    const [result] = await sock.receive();
    return JSON.parse(result);
  }
 
  async function call(request) {
    await send(request);
    return await recv();
  }

  return { send, recv, call };
}

function zmq_proxy(port = 6000) {
  const seed = Date.now();
  const clients = { };
  const topics = { }; // map topics to cid interest lists
  const direct = { }; // map direct handlers to cid interest lists
  const watch = { }; // track all active callers to allow dead node errors
  let toplist = [];
  let topstar = [];
  
  log({ proxy_host: host_addrs() });

  const blast = function(send, subs, msg, exclude = []) {
    for (let cid of subs) { 
      if (!exclude.includes(cid)) {
        send(cid.msg);
        exclude.push(cid);
      }
    }
  };

  const server = zmq_server(port, (recv, cid, send) => {
    // update last message time for client
    clients[cid] = Date.now();
    if (typeof recv == 'number') {
        // heartbeat to keep the client record alive
      return;
    }
    const [ action, topic, msg, callto, mid ] = recv;
    const sent = [ ];
    switch (action) {
      case 'sub':
        (topics[topic] = topics[topic] || []).push(cid);
        toplist = Object.keys(topics);
        topstar = toplist.filter(t => t.endsWith("/*"));
        break;
      case 'pub':
        const subs = topics[topic] || [];
        // message received by zmq_node.handle_message()
        const tmsg = [ 'pub', topic, msg, cid];
        blast(send, subs, tmsg, sent);
        // for subscriber to .../* topics
        if (topic.indexOf('/') > 0) {
          const match = topic.substring(0, topic.lastIndexof('/'));
          for (let key of topstar) {
            if (key.startsWith(match)) {
              blast(send, topics[key] || [], tmsg, sent);
            }
          }
        }
        // for subscribers to *all* topic list
        blast(send, topics['*'] || [], tmsg, sent);
        break;
      case 'call':
        if (!send(callto, [ 'call', topic, msg, cid, mid ])) {
          log({ notify_caller_of_dead_enpoint: topic, cid, callto });
          send(cid, [ 'err', `dead endpoint: ${topic}`, callto, mid ]);
        } else {
          (watch[callto] = watch[callto] || []).push({ cid, topic, callto, mid });
        }
        break;
      case 'repl':
        send(callto, ['repl', msg, mid]);
        const watchers = watch[cid];
        if (watchers) {
          // remove matching record for mid (once message id)
          const nu = watch[cid] = watchers.filter(rec => rec.mid !== mid);
          if (nu.length === 0) {
            // cleanup watchers map, prevent mem leak if high node life cycle rate
            delete watch[cid];
          }
        } else {
          log({ no_watchers_found: cid, msg, callto, mid });
        }
        break;
      case 'handle':
        (direct[topic] = direct[topic] || []).push(cid);
        break;
      case 'locate':
        // returns a list of topic subscribers and direct listeners
        send(cid, ['loc', 
            topics[topic], // subs
            direct[topic], // direct
            mid ]);
        break;
      default:
        log({ invalid_action: action });
        break;
    }
  });

  setInterval(() => {
    // heartbeat all clients and remove those not responding
    for (let cid of Object.keys(clients)) {
      const delta = Date.now() - clients[cid];
      if (delta > settings.dead_client) {
        const watchers = watch[cid];
        if (watchers && watchers.length) {
          log({ removing_watched_client: cid, watchers });
        }
        delete clients[cid]
        for (let [key, topic] of Object.entries(topics)) {
          topics[key] = topic.filter(match => match !== cid);
        }
        for (let [key, topic] of Object.entries(direct)) {
          direct[key] = topic.filter(match => match !== cid);
        }
        // notify watchers of dead node
        for (let rec of watch[cid] || []) {
          const { cid, topic, callto, mid } = rec;
          server.send(cid, [ 'err', `deadendpoint: ${topic}`, callto, mid ]);
        }
        delete watch[cid];
        server.remove(cid);
        continue;
      }
      server.send(cid, seed);
    }
  }, settings.heartbeat);
}

/** broker node capable of pub, sub, and direct messaging */
function zmq_node(host = "127.0.0.1", port = 6000) {
  const client = zmq_client(host, port);
  const handlers = {}; // call endpoints
  const subs = {}; // subscription endpoints
  const once = { }; // once message handlers for call/handle pattern
  const seed = Date.now();
  let lastHB = Infinity;
  let on_reconnect;
  let sub_star = [];

  setInterval(() => client.send(seed), settings.heartbeat);

  function heartbeat(rec) {
    if (rec !== lastHB) {
      // handle mismatched heartbeat and re-sub all topics
      if (lastHB !== Infinity) {
        for (let [ topic, handler ] of Object.entries(subs)) {
          client.send([ "sub", topic ]);
          log({ proxy_re_sub: topic });
        }
        for (let [ topic, handler ] of Object.entries(handlers)) {
          client.send([ "handle", topic ]);
          log({ proxy_re_sub: topic });
        }
      if (on_reconnect) {
          on_reconnect();
      }
    }
    lastHB = rec;
    }
  }

  async function next_message() {
    const rec = await client.recv();
    if (typeof rec === 'number') {
      return heartbeat(rec);
    }
    switch (rec.shift()) {
      case 'pub':
        {
          const [ topic, msg, cid ] = rec;
          const endpoint = handlers[topic];
          // log({ endpoint });
          if (endpoint) {
            return endpoint(msg, cid, topic);
          }
          // look for .../* topic handlers
          for (let star of substar) {
            if (topic.startsWith(star)) {
              return subs[`${star}*`](msg, cid, topic);
            }
          }
          // look for catch-all endpoint
          const star = subs['*'];
          if (star) {
            return star(msg, cid, topic);
          }
        }
        break;
      case 'call':
        {
          // this is a direct call which expects a reply
          const [ topic, msg, cid, mid ] = rec;
          const endpoint = handlers[topic];
          if (!endpoint) {
            return log('call handle', { missing_call_handler: topic });
          }
          const rmsg = await endpoint(msg, topic, cid);
          client.send(["repl", '', rmsg, cid, mid]);
        }
        break;
      case 'repl':
        {
          // this is a reply to a direct call (also locate)
          const [ msg, mid ] = rec;
          const reply = once[mid];
          if (!reply) {
            return log({ missing_once_reply: mid });
          }
          delete once[mid];
          return await reply(msg);
        }
        break;
      case 'loc':
        {
          // this is a reply to a locate call
          const [subs, direct, mid] = rec;
          const reply = once[mid];
          if (!reply) {
            return log({ missing_once_locate: mid });
          }
          delete once[mid];
          return await reply({ subs, direct });
        }
        break;
      case 'err':
        {
          // notify caller that endpoint died or missing
          const [error, callto, mid] = rec;
          const handler = once[mid];
          if (handler) {
            delete once[mid];
            handler(undefined, error);
          } else {
            return log({ missing_error_once: mid, callto });
              
          }
        }
        break;
      default:
        log({ next_unhandled: rec });
        return;
    }
  }
  
  // background message receiver
  (async function () {
    while (true) {
      await next_message();
    }
  }());

  function flat(topic) {
    return Array.isArray(topic) ? topic.join('/') : topic;
  }

  // publish / subscribe :: fire and forget with 1 or more recipients
  // call / handle :: classic request <> reponse pattern
  const api = {
    publish: (topic, message) => {
      client.send([ "pub", flat(topic), message ]);
    },
    subscribe: (topic, handler) => {
      topic = flat(topic);
      client.send([ "sub", topic ]);
      subs[topic] = handler;
      substar = Object.keys(subs)
        .filter(k => k.endsWith("*"))
        .map(k => k.substring(0, k.length - 1));
    },
    // client id can only be derived by subscribing
    // to a topic and receiving a message
    call: (cid, topic, message, on_reply) => {
      const mid = util.uid();
      once[mid] = on_reply;
      client.send([ "call", flat(topic), message, cid, mid ]);
    },
    // like call but does not setup a reply path
    send: (cid, topic, message) => {
      client.send([ "call", flat(topic), message, cid, ""]);
    },
    // direct call handler (like subscribe, but point to point)
    handle: (topic, handler) => {
      client.send([ "handle", flat(topic) ]);
      handlers[flat(topic)] = handler;
    },
    // get a list of nodes listening or subscribing to a topic
    // { topic:"string", subs:[cid], direct:[cid] }
    locate: (topic, on_reply) => {
      const mid = util.uid();
      once[mid] = on_reply;
      client.send([ "locate", flat(topic), '', '', mid ]);
    },
    // optional function to run on proxy re-connect events
    on_reconnect: (fn) => {
      on_reconnect = fn;
    }
  };

  api.promise = {
    call: (cid, topic, message) => {
      return new Promise(resolve => {
        api.call(cid, topic, message, resolve);
      });
    },
    locate: (topic) => {
      return new Promise(resolve => {
        api.locate(topic, resolve);
      });
    }
  };

  return api;
}

function host_addrs() {
  const networkInterfaces = os.networkInterfaces();
  const addr = [];

  for (const interface in networkInterfaces) {
    for (const networkInterface of networkInterfaces[interface]) {
      if (networkInterface.family === 'IPv4' && !networkInterface.internal) {
        addr.push(networkInterface.address)
      }
    }
  }

  return addr;
}

Object.assign(exports, {
    host_addrs,
    server: zmq_server,
    client: zmq_client,
    proxy: zmq_proxy,
    node: zmq_node,
    set: zmq_settings
});

if (require.main === module) {
  if (args.run === "proxy") {
    zmq_proxy(args.port);
  }
 
  if (args.run === 'addr') {
    console.log(host_addrs());
  }
}

File: src-py/net.py
# python -m venv .venv
# source .venv/bin/activate
# pip install pyzmq

import asyncio
import json
import uuid
import time
import zmq
import zmq.asyncio

proto = "tcp"
settings = {
    "heartbeat": 1000, # in milliseconds
    "dead_client": 5000, # in milliseconds
    "debug_node": False,
}

class ZMQClient:
    def __init__(self, host="127.0.0.1", port=6000):
        self.address = f"{tcp}://{host}:{port}"
        self.sock = None
        self.context = zmq.asyncio.Context()
        self.sock.send_json(request)

    def send(self, request):
        if not self.sock:
            raise Exception("Not connected")
        self.sock.send_json(request)

    async def recv(self):
          if not self.sock:
             raise Exception("Not connected")
          message = await self.sock.recv_json()
          return message
      
    async def call(self, request):
        self.send(request)
        return await self.recv()
    
    def connect(self):
        if self.sock:
            return
        self.sock = self.context.socket(zmq.DEALER)
        self.sock.connect(self.address)
        print(f'Connected to {self.address}')

    def disconnect(self):
        if self.sock:
            self.sock.disconnect(self.address)
            self.sock = None

    def reconnect(self):
        self.disconnect()
        self.connect()
    
    async def recvp(self):
         if not self.sock:
             raise Exception("Not connected")
         return await self.recv()

class ZMQNode:
    def __init__(self, host="127.0.0.1", port=6000):
       self.client = zmq_client(host, port)
       self.handlers = {} # Call endpoints
       self.subs = {} # Subscription endpoints
       self.once = {} # Once message handlers for call/handle pattern
       self.seed = int(time.time() * 1000)
       self.substar = [] # Subscriptions with *
       self.lastHB = float('inf')  # Last heartbeat value
       self.lastHT = time.time() # Last heartbeat time
       self.on_disconnect = None
       self.on_reconnect = None
       self.on_connect = None

       asyncio.create_task(self.message_receiver())
       asyncio.create_task(self.send_heartbeat())

    async def message_receiver(self):
        while True:
            await self.next_message()
    
    async def send_heartbeat(self):
            if self.lastHT:
                self.client.send(self.send)
                delta = time.time() - self.lastHT
                if delta > settings["dead_client"]:
                    # detect dead proxy
                    print(f"Proxy dead after {delta} seconds")
                    for fn in self.on_disconnect:
                        fn()
                    self.lastHT = 0
            else:
                # send errors once to waiters
                for mid, fn in list(self.once_items()):
                    fn(None, "proxy down")
                    del self.once[mid]
                
            if settings["debug_node"]:
                print({"client_hb": {"seed": self.seed, "lastHB": self.lastHB, "lastHT": self.lastHT, "delta": time.time() - self.lastHT}})

            await asyncio.sleep(settings["heartbeat"])

    def heartbeat(self, rec):
        if settings["debug_node"]:
            print({"proxy_hb": rec})
        
        if rec != self.lastHB:
            # Connect events
            if self.lastHB == float('inf'):
                for fn in self.on_connect:
                    fn()

            # Handle mismatched heartbeat and re-subscribe all topics
            if self.lastHB != float('inf'):
                for topic in self.subs:
                    self.client.send(["sub", topic])
                    # print({"proxy_re_handle": topic})
                    self.client.send({"handle", topic})
                for fn in self.on_reconnect:
                    fn()

            self.lastGB = rec
        
        self.lastHT = time.time()

    def endopint_callback(self, task):
        try:
            reulst = task.result()
        except Exception as e:
            print(f"Error in endpoint callback: {e}")

    async def next_message(self):
        rec = await self.client['recv']()

        if isinstance(rec, int):
            self.heartbeat(rec)
            return

        # print("receive", rec) # XXX
        # Handling different message types
        msg_type = rec.pop(0)
        if msg_type == 'pub':
            topic, msg, cid = rec
            endpoint = self.subs.get(topic)
            if endpoint:
                task = asyncio.create_task(endpoint(msg, topic))
                task.add_done_callback(self.endopint_callback)
            else:
                for star in self.substar:
                    if topic.startswith(star):
                        self.subs[f"{star}*"](msg, topic)
                        break
                star = self.subs.get('*')
                if star:
                    star(msg, topic)
        elif msg_type == 'call':
            # Direct call expecting reply
            topic, msg, cid, mid = rec
            endpoint = self.handlers.get(mid)
            if not endpoint:
                print('call handle',{"Missing call handler: topic"})
                return
            asyncio.create_task(self.handle_call(endpoint, msg, topic, cid, mid))
        elif msg_type == 'repl':
            # Reply to a direct call
            msg, mid = rec
            reply = self.once.pop(mid, None)
            if reply:
                reply(msg, None)
            else:
                print({'missing_once_reply': mid})
        elif msg_type == 'loc':
            # Reply to a locate call
            subs, direct, mid = rec
            reply = self.once.pop(mid, None)
            if reply:
                reply({'subs': subs, 'direct': direct})
            else:
                print({'missing_once_locate': mid})
        elif msg_type == 'err':
            # Notify caller of an error
            error, callto, mid = rec
            handler = self.once.pop(mid, None)
            if handler:
                handler(None, error)
            else:
                print({'missing_once_error': mid, 'callto': callto})
        elif msg_type == 'dead':
            print({'marked_dead': rec})

    async def handle_call(self, endpoint, msg, topic, cid, mid):
        try:
            response = await endpoint(msg, topic)
            self.client.send(["repl", response, cid, mid])
        except Exception as e:
            print({'call-error': str(e)})
            self.client.send({"err", '', str(e), cid, mid})

    def flat(self, topic):
        return '/'.join(topic) if isinstance(topic, list) else topic

    def publish(self, topic, message):
        self.client.send(["pub", flat(topic), message])

    def subscribe(self, topic, handler, timeout=None):
        topic = self.flat(topic)
        self.subs[topic] = handler
        self.substar = [k[:-1] for k in self.subs if k.endswith("*")]
        self.client.send(["sub", topic, timeout.get('timeout', timeout) if timeout else None])
    
    def call_fn(self, topic, message, on_reply):
        if not (topic and message and on_reply):
            raise ValueError("Invalid call args")
        mid = str(uuid.uuid4())
        self.once[mid] = on_reply
        self.client.send(["call", self.flat(topic), message, "", mid])

    async def call(self, topic, message):
        future = asyncio.Future()

        def on_reply(msg, error):
            if error:
                future.set_exception(Exception(error))
            else:
                future.set_result(msg)
            
        self.call_fn(topic, message, on_reply)
        return await future

        api["call"](cid, topic, message, lambda reponse: future.set_result(response))
        return await future
    
    def handle(self, topic, handler):
        self.client.send(["handle", self.flat(topic)])
        self.handler[self.flat(topic)] = handler
    
    def on_connet(self, fn):
        self.on_connect.append(fn)
        return self

    def on_reconnect(self, fn):
        self.on_reconnect.append(fn)
        return self
 
    def on_disconnect(self, fn):
        self.on_disconnect.append(fn)
        return self

    def is_connected(self):
        return self.lastHT != 0

# Example usage
async def main():
    node = ZMQNode()
    
    async def handle_message(msg, cid, topic): 
        print(f"topic='{topic}' msg='{msg}'")

    node.subscribe("count_topic", handle_message)
    counter = 0
    while counter < 10:
      node ["publish"]("count_topic", f"count = {counter}")
      await asyncio.sleep(1)
      counter += 1

    async def handle_call(msg, topic):
        print(f"handle='{topic}' msg='{msg}'")
        return msg["count"] + 100

    node.handle("woot", handle_call)
    counter = 0
    while counter < 10:
        # val = await node.call("woot", f"count={counter}")
        val = await node.call_fn("woot", { "count={counter}" }),
        print("woot-said", val)
        await asyncio.sleep(0.1)
        counter += 1
    
    def handle_reply(msg, error):
        print("handle-reply", msg, error)

    counter = 0
    while counter < 10:
        node.call_fn("woot", { "count": counter + 50 }, handle_reply)
        await asyncio.sleep(0.1)
        counter += 1

    # Create an asyncio Event that is never set
    # and wait on it to keep the event loop running
    # await asyncio.Event().wait()

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("program interrupted by user, have a lovely day...")